domain/application/service/service.go

```diff
@@ -445,6 +445,80 @@ func (s *WatchableService) WatchApplicationConfigHash(ctx context.Context, name
 	)
 }
 
+// WatchUnitAddressesHash watches for changes to the specified unit's
+// addresses hash, as well as changes to the endpoint bindings for the spaces
+// the addresses belong to.
+//
+// If the unit does not exist an error satisfying [applicationerrors.UnitNotFound]
+// will be returned.
+func (s *WatchableService) WatchUnitAddressesHash(ctx context.Context, unitName coreunit.Name) (watcher.StringsWatcher, error) {
+	appUUID, err := s.st.GetApplicationIDByUnitName(ctx, unitName)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	// currentHash is the current hash. This will be filled in by the initial
+	// query.
+	// If it's empty after the initial query, then a new address hash will be
+	// generated on the first change.
+	var currentHash string
+
+	// Retrieve the net node uuid that corresponds to the cloud service and if
+	// there isn't one, then the unit's net node.
+	netNodeUUID, err := s.st.GetNetNodeUUIDByUnitName(ctx, unitName)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	ipAddressTable, appEndpointTable, query := s.st.InitialWatchStatementUnitAddressesHash(appUUID, netNodeUUID)
+	return s.watcherFactory.NewNamespaceMapperWatcher(
+		func(ctx context.Context, txn database.TxnRunner) ([]string, error) {
+			initialResults, err := query(ctx, txn)
+			if err != nil {
+				return nil, errors.Capture(err)
+			}
+
+			if num := len(initialResults); num > 1 {
+				return nil, errors.Errorf("too many address hashes for unit %q", unitName)
+			} else if num == 1 {
+				currentHash = initialResults[0]
+			}
+
+			return initialResults, nil
+		},
+		func(ctx context.Context, _ database.TxnRunner, changes []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+			// If there are no changes, return no changes.
+			if len(changes) == 0 {
+				return nil, nil
+			}
+
+			newHash, err := s.st.GetAddressesHash(ctx, appUUID, netNodeUUID)
```

Comment:
As a model gets larger this is going to get a lot of action for units that are not affected by one change or another, for example, the deployment of a new application, which would fire the app endpoints subscription, causing everyone to re-query their hashes.

Make a note of this behaviour in a comment, with the possible future action being to accumulate the change events, and check whether this unit is effected before fetching and recalculating.

Reply:
Sure thing, this is the note I added in the method doc:
```
//
// NOTE(nvinuesa): This method is used in the `WatchUnitAddressesHash` watcher
// to validate if a change has indeed occured. The issue with this behavior is
// that it will get fired very often and the probability of a change that is
// of interest for the unit is low.
// A possible future improvement would be to accumulate the change events and
// check whether the unit of interest has been affaceted, before hitting the db.
```